proto/backend.proto

```diff
@@ -0,0 +1,212 @@
+  map<string,string> decryptedSecureJsonData = 7;
+}
+
+
+//-----------------------------------------------
+// All Backend Plugins Implement
+//-----------------------------------------------
+
+service Core {
+  // Get the plugin status
+  rpc Check(PluginStatusRequest) returns (PluginStatusResponse);
+
+  // HTTP Style request
+  rpc Resource(ResourceRequest) returns (ResourceResponse);
+
+  // Well typed query interface
+  rpc DataQuery(DataQueryRequest) returns (DataQueryResponse);
+
+}
+
+//-----------------------------------------------
+// Plugin Status Check
+//-----------------------------------------------
+
+message PluginStatusRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  // Environment info
+  map<string,string> headers = 2;
+
+  enum FetchInfo {
+    STATUS = 0;  // nothing
+    API = 1;     // OpenAPI description
+    METRICS = 2; // Prometheus style metrics
+    DEBUG = 3;   // JSON Debug Info (admin+ view)
+  }
+
+  // Get an OpenAPI description (viewer+)
+  FetchInfo fetch = 3;
+}
+
+message PluginStatusResponse {
+  enum PluginStatus {
+    UNKNOWN = 0; // starting etc
+    OK = 1;
+    ERROR = 2;
+  }
+
+  PluginStatus status = 1;
+  string info = 2;
+}
+
+//-----------------------------------------------
+// Data Query
+//-----------------------------------------------
+
+message TimeRange {
+  int64 fromEpochMS = 1;
+  int64 toEpochMS = 2;
+}
+
+message DataQuery {
+  string refId = 1;
+  int64 maxDataPoints = 2;
+  int64 intervalMS = 3;
+  TimeRange timeRange = 4; // NOTE: this moved from QueryRequest in v1
+  bytes json = 5; // Arbitrary structure (model?  modelJson in v1)
+}
+
+message DataQueryRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  // Environment info
+  map<string,string> headers = 2;
+
+  // List of queries
+  repeated DataQuery queries = 3;
+}
+
+message DataQueryResponse {
+  // Arrow encoded DataFrames
+  // Each frame encodes its own: Errors, meta, and refId
+  repeated bytes frames = 1;
+
+  // Additional response metadata
+  map<string,string> metadata = 2;
+}
+
+//-----------------------------------------------
+// HTTP style request
+//-----------------------------------------------
+
+message ResourceRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  
+  // HTTP Style parameters
+  map<string,string> headers = 2; // Request Headers
+  string method = 3; // GET/POST/DELETE/etc
+  string path = 4;   // (with params?)
+  bytes body = 5;    // assume string unless header says otherwise?
+}
+
+message ResourceResponse {
+  int32 code = 1;  // Match HTTP status codes
+  map<string,string> headers = 2; // Response Headers
+
+  bytes body = 3;
+}
+
+
+//-----------------------------------------------
+// Allow plugins to access other plugins
+// Only allow this for some plugins?  GEL only
+//
+// ???? How does the backend know the PluginConfig????
+//-----------------------------------------------
+
+service GrafanaPlatform {
```

Comment:
As we did this in the previous iteration, these calls are grafana-server request bound (which is fine for transform, but may not be what we want).

So for example with, Transform - A request is made via HTTP api to Grafana-server, it makes a plugin request, calls backs are available during that request. The implementation here is the GrafanaPlatform Service is available during other requests.

However, if you wanted a plugin that could call things on a schedule, then the request would be initiated by the plugin. In which case these requests are not tied to the context of a request that originates from the server.